packages/grafana-ui/src/components/Logs/LogRowMessage.tsx

```diff
@@ -52,6 +54,44 @@ const getStyles = stylesFactory((theme: GrafanaTheme) => {
   };
 });
 
+const restructureLog = (line: string, prettifyLogMessage: boolean): string => {
+  if (prettifyLogMessage) {
+    try {
+      const parser = getParser(line) as LogsParser;
+      const fields = parser.getFields(line);
+      return fields.join('\n');
+    } catch (error) {
+      return line;
+    }
+  }
+  return line;
+};
+
+function renderLogMessage(
+  needsHighlighter: boolean | '' | undefined,
+  hasAnsi: boolean,
+  restructuredEntry: string,
+  highlights: string[] | undefined,
+  highlightClassName: string
+) {
+  if (needsHighlighter) {
+    return (
+      <Highlighter
+        textToHighlight={restructuredEntry}
+        searchWords={highlights ?? []}
+        findChunks={findHighlightChunksInText}
+        highlightClassName={highlightClassName}
+      />
+    );
+  } else if (hasAnsi) {
+    return <LogMessageAnsi value={restructuredEntry} />;
+  } else {
+    return restructuredEntry;
+  }
+}
+
+const memoizeRestructureLog = memoizeOne(restructureLog);
```

Comment:
You could memoize this directly in restructureLog - such as here in [sortLogs](https://github.com/grafana/grafana/blob/main/packages/grafana-ui/src/components/Logs/LogRows.tsx#L76-L78)

Reply:
Oh that's smart! By the way, is it a good idea to memoize the `renderLogMessage` function as well?

Reply:
Not necessarily. Memoization is useful when function is pure (which it is in both cases) and when we are doing heavier computing functions - e.g. figuring out parser and then parsing each log line that can be long (which we do in `restrucutreLog`). `renderLogMessage` is pretty simple and just returns components based on arguments. 

So the memoization in `restructureLog` is useful so we don't have to do whole parsing every time we toggle context or change wrapLogMessage. or add a text to highlight.